aten/src/ATen/core/op_registration/op_registration.cpp

```diff
@@ -165,4 +165,54 @@ RegisterOperators::~RegisterOperators() = default;
 RegisterOperators::RegisterOperators(RegisterOperators&&) noexcept = default;
 RegisterOperators& RegisterOperators::operator=(RegisterOperators&&) noexcept = default;
 
+
+CppFunction::CppFunction(KernelFunction func, std::unique_ptr<c10::FunctionSchema> schema)
+  : func_(std::move(func))
+  , schema_(std::move(schema))
+  {}
+
+Module::Module(const char* ns)
+  : ns_(ns)
+  {}
+
+Module::Module(Module&&) noexcept = default;
+Module& Module::operator=(Module&&) noexcept = default;
+
+
+Module&& Module::def(const char* schema) && {
+  register_.op(c10::RegisterOperators::options()
+    .schema(schema)
+    .aliasAnalysis(c10::AliasAnalysisKind::FROM_SCHEMA));
+  return std::move(*this);
+}
+
+static std::string addNamespace(const char* ns, const char* unqual_name_or_schema) {
+  if (ns) {
+    // TODO: slow!  Fix internal data structures so I don't have to paste the
+    // names together
+    std::ostringstream oss;
+    oss << ns << "::" << unqual_name_or_schema;
+    return oss.str();
+  } else {
+    return unqual_name_or_schema;
+  }
+}
+
+Module&& Module::def(const char* unqual_name, CppFunction&& f) && {
```

Comment:
Basil just noticed that the argument here should be called "schema_or_unqual_name" instead of "unqual_name"

Reply:
Well, if this is truly `schema_or_unqual_name` then our prepending of namespaces is inconsistent - in `def(schema)` we're passing it through unmodified.

Reply:
Adding this as a review so the preceding doesn't get lost - seems like we need to either
a) handle names and schemas uniformly w.r.t. prepending namespaces, or 
b) snoop the string to decide whether it's a name or schema, or
c) define this version of `def` as taking unqualified names and never schemas as its first argument

Reply:
Yes there's a bug. It wasn't caught in the integration tests because the `impl()` all have full schema so you end up with the "right" thing in the end.